lib/phoenix_live_view/tag_engine.ex

```diff
@@ -1160,6 +1163,93 @@ defmodule Phoenix.LiveView.TagEngine do
     end
   end
 
+  defp handle_keyed_comprehension(key_expr, state, tag_meta) do
+    # for now, we only support plain tags because we don't know if a function component
+    # renders to a single tag which is required by live components
+    case state.tag_handler.classify_type(tag_meta.tag_name) do
+      {:tag, _} ->
+        :ok
+
+      other ->
+        type =
+          case other do
+            {:local_component, _} -> "function components"
+            {:remote_component, _} -> "function components"
+            # special attrs for slots are handled separately, see wrap_special_slot
+            other -> inspect(other)
+          end
+
+        raise_syntax_error!(
+          "keyed :for comprehensions are not supported on #{type}",
+          tag_meta,
+          state
+        )
+    end
+
+    for_expr =
+      tag_meta[:for] || raise_syntax_error!("cannot use :key without :for", tag_meta, state)
+
+    if_expr = tag_meta[:if]
+    for_assigns = extract_variables(for_expr)
+    vars_changed = Enum.map(for_assigns, fn {name, _} -> name end)
+
+    ast =
+      invoke_subengine(state, :handle_body, [
+        [skip_require: true, root: true, vars_changed: vars_changed]
+      ])
```

Comment:
This feels semantically wrong. `handle_body` is meant to call only when we have finished everything. Could we pass these additional values to `handle_end` instead?

Reply:
If we do this, we may be able to avoid the push/pop dance downstream.

Reply:
If we don't use handle_body, we need to change the engine to turn the special render function to a rendered struct somehow